stl/inc/type_traits

```diff
@@ -1671,17 +1671,29 @@ struct _Invoke_traits_zero {
 template <class _Callable>
 using _Decltype_invoke_zero = decltype(_STD declval<_Callable>()());
 
+template <class _Void, class _Rx, class _Callable, class... _Args>
+struct _Invocable_r_nonvoid : false_type {};
+
+template <class _Rx, class _Callable, class... _Args>
+struct _Invocable_r_nonvoid<
+    void_t<decltype(_Implicitly_convert_to<_Rx>(_STD invoke(_STD declval<_Callable>(), _STD declval<_Args>()...)))>,
+    _Rx, _Callable, _Args...> : true_type {};
+
+template <class _Rx, class _Callable, class... _Args>
+struct _Is_nothrow_invocable_r_nonvoid : bool_constant<noexcept(_Implicitly_convert_to<_Rx>(
+                                             _STD invoke(_STD declval<_Callable>(), _STD declval<_Args>()...)))> {};
+
 template <class _Callable>
 struct _Invoke_traits_zero<void_t<_Decltype_invoke_zero<_Callable>>, _Callable> {
     // selected when _Callable is callable with zero _Args
     using type                  = _Decltype_invoke_zero<_Callable>;
     using _Is_invocable         = true_type;
     using _Is_nothrow_invocable = bool_constant<noexcept(_STD declval<_Callable>()())>;
     template <class _Rx>
-    using _Is_invocable_r = bool_constant<disjunction_v<is_void<_Rx>, is_convertible<type, _Rx>>>;
+    using _Is_invocable_r = bool_constant<disjunction_v<is_void<_Rx>, _Invocable_r_nonvoid<void, _Rx, _Callable>>>;
     template <class _Rx>
-    using _Is_nothrow_invocable_r = bool_constant<
-        conjunction_v<_Is_nothrow_invocable, disjunction<is_void<_Rx>, _Is_nothrow_convertible<type, _Rx>>>>;
+    using _Is_nothrow_invocable_r = bool_constant<conjunction_v<_Is_invocable_r<_Rx>,
+        disjunction<is_void<_Rx>, _Is_nothrow_invocable_r_nonvoid<_Rx, _Callable>>>>;

```

Comment:
This doesn't appear to be correct. Previously, we asked "are you nothrow-invocable, *and* (are we returning `void` in which case there's nothing more to check, or is the return type conversion non-throwing)?". Now, it appears that asking `is_nothrow_invocable_r<void, TotallyThrowsExceptions>` will be implemented with "are you **invocable** with this return type, *and* (is the return type `void`, or new fancy check)?". Since `TotallyThrowsExceptions` is invocable (but not nothrow-invocable), the answer has changed. Ditto for the nonzero `_Args` below.

Am I totally missing something? If this is a real issue, it additionally indicates missing test coverage (my fault! :joy_cat: I must have been thinking "well, we'll always check `is_nothrow_invocable` before considering the return type conversion, I don't need to repeat any of that coverage").

Reply:
Ouch, you are 100% correct. (About the bug, and about the test coverage.)